client/shared/src/components/FileMatchChildren.tsx

```diff
@@ -67,20 +67,48 @@ export const FileMatchChildren: React.FunctionComponent<FileMatchProps> = props
         context,
     ])
 
-    const { result, isLightTheme, fetchHighlightedFileLines } = props
+    // If optimizeHighlighting is enabled, compile a list of the highlighted file ranges we want to
+    // fetch (instead of the entire file.)
+    const optimizeHighlighting =
+        props.settingsCascade.final &&
+        !isErrorLike(props.settingsCascade.final) &&
+        props.settingsCascade.final.experimentalFeatures &&
+        props.settingsCascade.final.experimentalFeatures.enableFastResultLoading
+
+    const highlightRanges = !optimizeHighlighting
+        ? undefined
+        : grouped.map(
+              (group): IHighlightLineRange => ({
+                  startLine: group.startLine,
+                  endLine: group.endLine,
+              })
+          )
+
+    const { result, isLightTheme, fetchHighlightedFileLineRanges } = props
     const fetchHighlightedFileRangeLines = React.useCallback(
         (startLine, endLine) =>
-            fetchHighlightedFileLines(
+            fetchHighlightedFileLineRanges(
                 {
                     repoName: result.repository.name,
                     commitID: result.file.commit.oid,
                     filePath: result.file.path,
                     disableTimeout: false,
                     isLightTheme,
+                    ranges: highlightRanges,
                 },
                 false
-            ).pipe(map(lines => lines.slice(startLine, endLine))),
-        [result, isLightTheme, fetchHighlightedFileLines]
+            ).pipe(
+                map(lines =>
+                    !optimizeHighlighting
+                        ? lines[0].slice(startLine, endLine)
+                        : lines[
+                              grouped.findIndex(
+                                  group => group.startLine === startLine && group.endLine === group.endLine
```

Comment:
RHS of `&&` always true, probably:

```suggestion
                                  group => group.startLine === startLine && group.endLine === endLine
```